torch/distributed/_tensor/experimental/attention.py

```diff
@@ -231,6 +238,86 @@ def __call__(
         ...
 
 
+class _RingRotater(ABC):
+    @abstractmethod
+    def __init__(self, pg: dist.ProcessGroup, seq_dim: int) -> None:
+        ...
+
+    @abstractmethod
+    def rotate(self, buffer: torch.Tensor, curr_idx: int) -> torch.Tensor:
+        ...
+
+    @abstractmethod
+    def maybe_wait(self, tensor: torch.Tensor) -> torch.Tensor:
+        ...
+
+
+class _AllToAllRotater(_RingRotater):
+    """Use all_to_all to send the kv to the next rank"""
+    def __init__(self, pg: dist.ProcessGroup, seq_dim: int) -> None:
+        self._pg = pg
+        self._seq_dim = seq_dim
+
+    def rotate(self, buffer: torch.Tensor, curr_idx: int) -> torch.Tensor:
+        buffer = buffer.contiguous()
+        size = dist.get_world_size(self._pg)
+        dsts = list(range(1, size)) + [0]
+        return ft_c.permute_tensor(buffer, dsts, self._pg)
+
+    def maybe_wait(self, tensor: torch.Tensor) -> torch.Tensor:
+        if isinstance(tensor, ft_c.AsyncCollectiveTensor):
+            return tensor.wait()
+        return tensor
+
+
+class _AllGatherRotater(_RingRotater):
+    """
+    Allgather the kv and return the only the requried kv.
+    Only one communication will be done.
+    """
+    def __init__(self, pg: dist.ProcessGroup, seq_dim: int) -> None:
+        self._pg = pg
+        self._seq_dim = seq_dim
+        self._buffer: Optional[torch.Tensor] = None
+        self._idx = 0
+
+    def rotate(self, buffer: torch.Tensor, curr_idx: int) -> torch.Tensor:
+        self._idx = curr_idx
+        if self._buffer is None:
```

Comment:
this code is actually quite confusing. 

it is strange that we have `self._buffer` and we accept `buffer` arg.  We seem to never assign anything to `self._buffer` and we seem to only use `buffer` but we use `self._buffer` to determine whether to perform the allgather and then we return self._buffer.

it makes me think that somewhere else, we are passing `self._buffer` as input to rotate, but that would be strange!